### PR TITLE
VLOG() to Text File

### PR DESCRIPTION
**Pull request by philipphack**
Redirects the output of all active `VLOG()` calls to a text file specified by the environment variable `TF_CPP_VLOG_FILENAME`. If the environment variable is not set, or if it points to a location that cannot be opened for writing, the output continues to `stderr`.

Attn: @reedwm.

### CODE REVIEW

File: tensorflow/core/platform/default/logging.h [hunk 1]
```diff
@@ -470,6 +470,24 @@ int64 MinLogLevelFromEnv();
 
 int64 MaxVLogLevelFromEnv();
 
+// A class for managing the text file to which VLOG output is written.
+// If the environment variable TF_CPP_VLOG_FILENAME is set, all VLOG
+// calls are redirected from stderr to a file with corresponding name.
+class VlogFileMgr {
```
Comment (reedwm) [inline tensorflow/core/platform/default/logging.h:476]: I think this class can be in the `.cc` file

File: tensorflow/core/platform/default/logging.cc [hunk 2]
```diff
@@ -534,9 +553,11 @@ void TFDefaultLogSink::Send(const TFLogEntry& entry) {
   }
 
   // TODO(jeff,sanjay): Replace this with something that logs through the env.
```
Comment (reedwm) [inline tensorflow/core/platform/default/logging.cc:555]: You can remove this TODO now

File: tensorflow/core/platform/default/logging.cc [hunk 3]
```diff
@@ -534,9 +553,11 @@ void TFDefaultLogSink::Send(const TFLogEntry& entry) {
...
-  fprintf(stderr, "%s.%06d: %c%s %s:%d] %s\n", time_buffer, micros_remainder,
-          sev, tid_buffer, entry.FName().c_str(), entry.Line(),
-          entry.ToString().c_str());
+  mutex mu;
+  mutex_lock l(mu);
```
Comment (reedwm) [inline tensorflow/core/platform/default/logging.cc:557]: This lock isn't needed. By the time this point is reached, the `VlogFileMgr` will be initialized, and fprintf is thread-safe.

File: tensorflow/core/platform/default/logging.cc [hunk 4]
```diff
@@ -494,6 +512,7 @@ void TFDefaultLogSink::Send(const TFLogEntry& entry) {
     abort();
   }
 #else   // PLATFORM_POSIX_ANDROID
+  static internal::VlogFileMgr vlog_file;
```
Comment (reedwm) [inline tensorflow/core/platform/default/logging.cc:515]: Make this const

**Review by reedwm — CHANGES_REQUESTED** [review]

**Comment by philipphack** [comment]
Thanks, Reed. I believe the thread safety of `fprintf ` is linked to POSIX compliance.

File: tensorflow/core/platform/default/logging.cc [hunk 5]
```diff
@@ -286,6 +286,42 @@ int64 MaxVLogLevelFromEnv() {
 #endif
 }
 
+// A class for managing the text file to which VLOG output is written.
+// If the environment variable TF_CPP_VLOG_FILENAME is set, all VLOG
+// calls are redirected from stderr to a file with corresponding name.
+class VlogFileMgr {
```
Comment (reedwm) [inline tensorflow/core/platform/default/logging.cc:292]: This can be put into the anonymous namespace up above

**Review by reedwm — CHANGES_REQUESTED** [review]
According to [this StackOverflow response](https://stackoverflow.com/a/41131253), fprintf is thread-safe since C11 (although it's ambiguous whether logs interleave). I don't think a lock is necessary.

**Review by reedwm — APPROVED** [review]

**Pull request merged**